include/pybind11/gil.h

```diff
@@ -199,3 +199,43 @@ class gil_scoped_release {
 PYBIND11_NAMESPACE_END(PYBIND11_NAMESPACE)
 
 #endif // !PYBIND11_SIMPLE_GIL_MANAGEMENT
+
+PYBIND11_NAMESPACE_BEGIN(PYBIND11_NAMESPACE)
+
+class scoped_critical_section {
+public:
+    /// This does not do anything if there's a GIL. On free-threaded Python,
+    /// it locks an object. This uses the CPython API, which has limits
+    scoped_critical_section(handle obj) : single(true) {
+#ifdef Py_GIL_DISABLED
+        PyCriticalSection_Begin(&section, obj.ptr());
+#endif
+    }
+
+    scoped_critical_section(handle obj1, handle obj2) : single(true) {
+#ifdef Py_GIL_DISABLED
+        PyCriticalSection2_Begin(&section2, obj1.ptr(), obj2.ptr());
+#endif
+    }
+
```

Comment:
Do we need these (copy-pasted-modified from the existing `scoped_*` types above)?

```
    scoped_critical_section(const scoped_critical_section &) = delete;
    scoped_critical_section &operator=(const scoped_critical_section &) = delete;
```

Reply:
Yes, those are required.